[AutoDiff] [API] Continue deriving conformances for `Differentiable`.

dan-zheng opened the pull request:
Continue implementing `Differentiable` derived conformances for struct types.
A follow-up to https://github.com/apple/swift/pull/21580.

This patch introduces support for:
- Synthesizing member `TangentVector` and `CotangentVector` structs.
  This is necessary when not all members have `Self == TangentVector ==
  CotangentVector`.
- Marking synthesized vector space structs and typealiases with
  `@_fieldwiseProductSpace`.
- Making synthesis consistent with recent changes to `Differentiable`.
  - #21629: Vector space associated types now conform to `AdditiveArithmetic`,
    not `VectorNumeric` (which is limited to a single `Scalar` type).
  - #21601: Use `@noDerivative` attribute to determine vector space struct
    members. A member in the original struct type marked with `@noDerivative`
    will not have corresponding members in vector space structs.
- Enabling derived conformances to `AdditiveArithmetic` and
  `VectorNumeric` for structs with no stored properties.
- Many bug fixes regarding code synthesis, details omitted.

There are some todos:
- Fix [SR-9595](https://bugs.swift.org/browse/SR-9595), which is related to mutually recursive associated types.
  The bug blocks `Differentiable` synthesis for structs whose members
  have generic types (e.g. `<T : Differentiable>`).
  Failing tests due to this bug are commented in
  `test/Sema/struct_differentiable.swift`.
- Synthesize `@differentiable(wrt: (self))` on stored properties of
   `Differentiable`-conforming types.
  - It's possible to do this synthesis during the derived conformances code
    path, but the derived conformances code path is not triggered if a type
    manually defines all `Differentiable` requirements.
- Add `Differentiable` derived conformances runtime tests.
- Future: enable `Differentiable` derived conformances for enums.

dan-zheng committed:
[API] Continue deriving conformances for `Differentiable`.

Continue implementing `Differentiable` derived conformances for struct types.
A follow-up to https://github.com/apple/swift/pull/21580.

This patch introduces support for:
- Synthesizing member `TangentVector` and `CotangentVector` structs.
  This is necessary when not all members have `Self == TangentVector ==
  CotangentVector`.
- Marking synthesized vector space structs and typealiases with
  `@_fieldwiseProductSpace`.
- Making synthesis consistent with recent changes to `Differentiable`.
  - Vector space associated types now conform to `AdditiveArithmetic`,
    not `VectorNumeric` (which is limited to a single `Scalar` type).
  - Use `@noDerivative` attribute to determine vector space struct
    members. A member in the original struct type marked with
    `@noDerivative` will not have corresponding members in vector space
    structs.
- Enabliing derived conformances to `AdditiveArithmetic` and
  `VectorNumeric` for structs with no stored properties.
- Many bug fixes regarding code synthesis, details omitted.

There are some todos:
- Fix SR-9595, which is related to mutually recursive associated types.
  The bug blocks `Differentiable` synthesis for structs whose members
  have generic types (e.g. `<T : Differentiable>`).
  Failing tests due to this bug are commented in
  `test/Sema/struct_differentiable.swift`.
- Synthesize `@differentiable(wrt: (self))` on stored properties of
   `Differentiable`-conforming types.
  - It's possible to do this synthesis during the derived conformances code
    path, but the derived conformances code path is not triggered if a type
    manually defines all `Differentiable` requirements.
- Add `Differentiable` derived conformances runtime tests.
- Future: enable `Differentiable` derived conformances for enums.

dan-zheng committed:
[API] Conform synthesized vector space structs to `KeyPathIterable`.

jekbradbury commented:
If a stored property doesn't conform to `Differentiable`, do we want to automatically make it `@noDerivative` or throw an error? Currently it errors, but the error is simply that the parent struct doesn't conform to `Differentiable`, suggesting that tangent vector synthesis didn't take place.

dan-zheng commented:
> If a stored property doesn't conform to `Differentiable`, do we want to automatically make it `@noDerivative` or throw an error? Currently it errors, but the error is simply that the parent struct doesn't conform to `Differentiable`, suggesting that tangent vector synthesis didn't take place.

That's a great question. For now, I think it's better to require explicit `@noDerivative` on stored properties because it makes it easier to answer "does a stored property have a corresponding vector space member"?

The condition for deriving conformances to `Differentiable` is basically: do all stored properties not marked with `@noDerivative` conform to `Differentiable`? <sub>(The check for `@noDerivative` isn't currently implemented, I'll add it now.)</sub>

Then, a stored property has a corresponding vector space member iff it is not marked `@noDerivative`. You can determine this at a glance.

---

Example:
```swift
struct BatchNorm : Differentiable {
  var offset: Tensor<Float> = Tensor(0)
  var scale: Tensor<Float> = Tensor(0)

  @noDerivative let axis: Int
  // Explicit `@noDerivative` above is very clear.

  // If `@noDerivative` was instead added implicitly, one would need to know
  // whether a stored property's type conforms to `Differentiable` or not to know
  // whether the stored property exists in associated vector spaces.
  // let axis: Int // I need to know whether `Int` conforms to `Differentiable` or not.

  @noDerivative let momentum: Tensor<Float> = Tensor(0.99)
  @noDerivative let epsilon: Tensor<Float> = Tensor(1e-3)
  @noDerivative var runningMean: Tensor<Float>
  @noDerivative var runningVariance: Tensor<Float>
}
```

rxwei commented:
> If a stored property doesn't conform to `Differentiable`, do we want to automatically make it `@noDerivative` or throw an error?

I'd say no, because I generally prefer explicit annotations.  We do certainly want synthesis-related diagnostics though, e.g. "property does not conform to `Differentiable`; do you want to add `@noDerivative`?"

dan-zheng committed:
Minor cleanup.

rxwei commented:
> Should synthesized `TangentVector` and `CotangentVector` conform to `KeyPathIterable`? My answer is definitely yes.

I'd prefer leaving it out of this patch. Having the compiler synthesize a `KeyPathIterable` conformance for a math protocol feels pretty arbitrary. Is there a compelling use case when that this solves? For optimizers that do not assume `TangentVector == CotangentVector == Self`, having these key paths won't help much because what we really need is a bijection between `Self`'s key paths and `TangentVector`/`CotangentVector`'s key paths for joint iteration. I think we should focus on solving the joint iteration problem (later, after this patch) instead of checking in a half solution.

dan-zheng committed:
[API] Un-conform synthesized vector space structs from `KeyPathIterable`.

Synthesizing `KeyPathIterable` for these structs is ad-hoc and limited in
usefulness (not useful when `Self != TangentVector` or `Self != CotangentVector`).

A more general solution involves providing a mapping between `Self`s
keypaths and vector space structs' keypaths.

rxwei reviewed (approved):
Great work!

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

dan-zheng committed:
Fix test/AutoDiff/noderivative-attr.swift.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
